src/Util/FisherYates/Implementation.dfy

```diff
@@ -0,0 +1,33 @@
+/*******************************************************************************
+ *  Copyright by the contributors to the Dafny Project
+ *  SPDX-License-Identifier: MIT
+ *******************************************************************************/
+
+module FisherYates.Implementation {
+  import Interface
+
+  trait {:termination false} Trait extends Interface.Trait {
+
+    method Shuffle<T(!new)>(a: array<T>)
+      decreases *
+      modifies this, a
+    {
+      var i := 0;
+      while i < a.Length {
+        var j := UniformIntervalSample(i, a.Length);
+        Swap(a, i, j);
```

Comment:
I don't trust the Dafny compiler to inline this, so perhaps we should do it by hand.